### PR TITLE
Liquidhandler.Plan: Speed improvements

### PR DESCRIPTION
**Pull request by haydnKing**
As part of https://github.com/antha-lang/antha/pull/901 (which added and centralised some tests on LH.Plan) I identified some pain points that could be fixed without affecting anything external.

a) wunit was quick sorting the list of valid units for each measurement type every time the list was requested - since some functions depend on them being in alphabetical order, they're now sorted once as they're added to the units registry

b) instruction generation was calling `inventory.NewTipBox` simply to access tip geometry, even though it had access to this information already

c) several slice operations in multitransferparams created a slice with an explicit length of 1 then added more than one thing to the slice - slices are now allocated with the right size

d) `colDup` (microArch/driver/transferblock.go) is improved, but remains the largest bottleneck (in these tests at least). Improving it further would involve a refactor of `get_parallel_sets` which seemed likely to have far reaching consequences.

Overall, these speed improvements knock a full run of `antha-tester` down from around 15 minutes to 10 minutes in "real" time

### CODE REVIEW

File: antha/anthalib/wunit/unitregistry.go [hunk 1]
```diff
@@ -116,13 +117,31 @@ func (self *UnitRegistry) declareUnit(measurementType string, unit *Unit) error
...
+
+	pos := 0
+	for pos < len(self.orderedUnitByType[measurementType]) {
+		if symbol <= self.orderedUnitByType[measurementType][pos] {
+			break
+		}
+		pos++
+	}
+
+	self.orderedUnitByType[measurementType] = append(self.orderedUnitByType[measurementType][:pos], append([]string{symbol}, self.orderedUnitByType[measurementType][pos:]...)...)
```
[the remark this answers is not in the document]
Comment (msackman-synthace) [reply antha/anthalib/wunit/unitregistry.go]: (err though this explodes if the list is empty so...)

[the remark this answers is not in the document]
Comment (msackman-synthace) [reply antha/anthalib/wunit/unitregistry.go]: (actually, it's fine provided the cap is never 0...)

File: microArch/driver/liquidhandling/transferblock.go [hunk 2]
```diff
@@ -441,6 +437,8 @@ func get_col(pdm wtype.Platedestmap, col, multi, wells int, contiguous, full boo
 		wellsPerTip = wells / multi
 	}
 
+	fmt.Printf("len(pdm[col])-2 = %d\n", len(pdm[col])-2)
```
[the remark this answers is not in the document]
Comment (haydnKing) [reply microArch/driver/liquidhandling/transferblock.go]: embarrassment

File: antha/anthalib/wunit/unit.go [hunk 3]
```diff
@@ -164,6 +164,12 @@ func (self *Unit) compatibleWith(pu PrefixedUnit) bool {
...
-	ret := *self
-	return &ret
+	return &Unit{
+		name:       self.name,
+		symbol:     self.symbol,
+		siSymbol:   self.siSymbol,
+		prefix:     self.prefix,
+		multiplier: self.multiplier,
+		exponent:   self.exponent,
+	}
```
[the remark this answers is not in the document]
Comment (msackman-synthace) [reply antha/anthalib/wunit/unit.go]: Interesting. I don't really mind; the only thing I'd say is that your change is fragile to being forgotten if a new field is added.

[the remark this answers is not in the document]
Comment (haydnKing) [reply antha/anthalib/wunit/unit.go]: yup, seems like a legit reason to favour the previous version to me

File: antha/anthalib/wunit/unitregistry.go [hunk 4]
```diff
@@ -116,13 +117,36 @@ func (self *UnitRegistry) declareUnit(measurementType string, unit *Unit) error
...
+	}
+
+	pos := 0
+	for ; pos < len(ordered); pos++ {
+		if symbol <= ordered[pos] {
+			break
+		}
+	}
+
+	ordered = append(ordered[:pos+1], ordered[pos:]...)
```
Comment (msackman-synthace) [inline antha/anthalib/wunit/unitregistry.go]: Wait, hang on, this isn't safe. If you reduce the capacity to a smaller number, say, 4, I expect this'll blow up. All you need to do is make sure you have space first in the array, so I'd do an `ordered = append(ordered, "")` first. That way the `:pos+1` should always be safe.

**Pull request merged**